### PR TITLE
Evergreen Client Registration and Authentication

### PR DESCRIPTION
**Pull request by rtyler**
This document is ready for a number and Draft status, see [this thread](https://groups.google.com/d/msgid/jenkinsci-dev/20180324020555.p26pfpln3j722m5r%40blackberry.coupleofllamas.com)

### CODE REVIEW

File: jep/0000/README.adoc [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+   |     <generate JWT enc token>     |                    |
+   |               |            HTTP 200 Ok                |
+   |               |         (respond with JWT)            |
+   |               o-------------------------------------->|
+   |               |                  |                    |
+   |               |                  |    Status request  |
+   |               |                  |      (with JWT)    |
+   |               |                  |<-------------------o
+   |               |                  |      HTTP 401      |
+   |               |                  o------------------->|
```
Comment (orrc) [inline jep/0000/README.adoc]: Presumably this should be HTTP 20x?

File: jep/0000/README.adoc [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+Clients in a registered state must log into the authentication service in order
+to access any other resources in the Evergreen hosted service layer. A `login`
+request consists of the UUID footnoteref:[uuid], signed by the client's private
+key. This provides the Authentication Service enough information to verify that
+the client is authentic.
+
+Once the client has been deemed authentic, the Authentication Service will
+generate a link:https://jwt.io[JSON Web Token] (JWT) which must be encrypted
+with a pre-shared key (PSK), known to the other backend services, but not the
+client. Additionally, JWT must contains a **14 day expiry** for the <<claims>>
```
Comment (orrc) [inline jep/0000/README.adoc]: contains -> contain

File: jep/0000/README.adoc [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+== Backwards Compatibility
+
+There is no previous "Jenkins Essentials registration system" and therefore no
+backwards compatibility concerns.
+
+
+== Security
+
+Securely registering and authenticating clients is the primary motivation and
+consequence of this desig. This section is intentionally empty as security
```
Comment (orrc) [inline jep/0000/README.adoc]: desig -> design

File: jep/0000/README.adoc [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+
+=== Certificate Pinning
+
+While this document expects that backend services are served via <<tls>>, it
+doesn't describe additional security enhancements which may come in a future
+design document such as:
+link:https://www.owasp.org/index.php/Certificate_and_Public_Key_Pinning[Certificate Pinning]
+
+Certificate pinning would ensure that clients are hard-coded to only trust the
+EVergreen backend service layer, further reducing the potential for
```
Comment (orrc) [inline jep/0000/README.adoc]: EVergreen -> Evergreen

File: jep/0000/README.adoc [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+container orchestration layer to deploy service containers in (e.g. the
+Registration service) and a PostgreSQL database for application storage.
+
+The specifics of these requirements will be defined in a future
+link:https://github.com/jenkins-infra/iep[Infrastructure Enhancement Proposal].
+
+
+== Testing
+
+Testing of these registration and authentication flow is performed within the
```
Comment (orrc) [inline jep/0000/README.adoc]: flow -> flows

**Comment by rtyler** [comment]
Thanks for spotting those errors @orrc, they've been fixed now

**Pull request merged**